internal/objects/reconciler.go

```diff
@@ -411,34 +412,55 @@ func (r *Reconciler) getTopSourceToPropagate(log logr.Logger, inst *unstructured
 	ns := r.Forest.Get(inst.GetNamespace())
 	// Get all the source objects with the same name in the ancestors excluding
 	// itself from top down.
-	nnms := ns.Parent().GetAncestorSourceNames(r.GVK, inst.GetName())
+	nnms := ns.Parent().GetAncestorSourceNamesWithLog(log, r.GVK, inst.GetName())
 	for _, nnm := range nnms {
 		// If the source cannot propagate, ignore it.
 		// TODO: add a webhook rule to prevent e.g. removing a source finalizer that
 		//  would cause overwriting the source objects in the descendents.
 		//  See https://github.com/kubernetes-sigs/multi-tenancy/issues/1120
+		log.Info("Searching for object in namspace", "namespace", ns.Name(), "nmnName", nnm.Name)
 		obj := r.Forest.Get(nnm.Namespace).GetSourceObject(r.GVK, nnm.Name)
 		if !r.shouldPropagateSource(log, obj, inst.GetNamespace()) {
+			log.Info("Found object but it should not be propegated ", "parentobject", obj.GetName(), "namespace", ns.Name())
 			continue
 		}
 		return obj
 	}
+
+	log.Info("Unable to find source object in parent", "namespace", ns.Name(), "namespaceParent", ns.Parent())
 	return nil
 }
 
 // syncPropagated will determine whether to delete the obsolete copy or overwrite it with the source.
 // Or do nothing if it remains the same as the source object.
-func (r *Reconciler) syncPropagated(inst, srcInst *unstructured.Unstructured) (syncAction, *unstructured.Unstructured) {
+func (r *Reconciler) syncPropagated(log logr.Logger, inst, srcInst *unstructured.Unstructured) (syncAction, *unstructured.Unstructured) {
 	ns := r.Forest.Get(inst.GetNamespace())
+
+	// Look up existing object this is used later to prevent us deleting an object
+	// that hnc is not yet aware of. srcObj should never be nil but log if we end up
+	// in this state
+	srcObj := ns.GetSourceObject(r.GVK, inst.GetName())
 	// Delete this local source object from the forest if it exists. (This could
 	// only happen when we are trying to overwrite a conflicting source).
-	ns.DeleteSourceObject(r.GVK, inst.GetName())
+	ns.DeleteSourceObject(log, r.GVK, inst.GetName())
 	stats.OverwriteObject(r.GVK)
 
 	// If no source object exists, delete this object. This can happen when the source was deleted by
 	// users or the admin decided this type should no longer be propagated.
 	if srcInst == nil {
-		return actionRemove, nil
+		log.Info("Unable to find srcInst removing object", "namespace", ns.Name(), "namespaceParent", ns.Parent(), "inst", inst.GetName())
+
+		if srcObj == nil {
```

Comment:
this does seem to get hit more frequently than I'd imagine, I think just because of cycles, but doesn't seem to cause any problems